File: PythonPractice/src/FindSubStringAndReturnCharIndexCountAsTuple.py
from typing import List, Tuple

# Define a tuple structure to store character, index, and count
class CharTuple:
    def __init__(self, character: str, index: int, count: int):
        self.character = character
        self.index = index
        self.count = count

    def __str__(self):
        return f"('{self.character}', {self.index}, {self.count})"

def find_substring_counter(input_string: str) -> List[CharTuple]:
    # Check for null or non-string inputs
    if input_string is None or not isinstance(input_string, str):
        raise ValueError("Pass a valid input string.")

    # Check for empty input
    if input_string.strip() == "":
        raise ValueError("Input string cannot be empty.")

    # Initialize the result list
    result: List[CharTuple] = []

    # If the input string has less than 2 characters, return an empty result
    if len(input_string) < 2:
        return result

    size = len(input_string)
    start = 0

    # Iterate through the string to find repeating characters
    while start < size:
        current_character = input_string[start]
        end = start

        while end < size and input_string[end] == current_character:
            end += 1

        count = end - start
        # If repeating characters are found, add to the result
        if count > 1:
            result.append(CharTuple(current_character, start, count))

        start = end

    return result

def main():
    input_string = "aabcdeee"

    try:
        subset_string = find_substring_counter(input_string)
        for char_tuple in subset_string:
            print(char_tuple)
    except ValueError as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()

File: JavaPractice/src/FindSubStringAndReturnCharIndexCountAsTuple.java
/*
Welcome to JDoodle!

You can execute code here in 88 languages. Right now you’re in the Java IDE.

  1. Click the orange Execute button ▶ to execute the sample code below and see how it works.

  2. Want help writing or debugging code? Type a query into JDroid on the right hand side ---------------->

  3.Try the menu buttons on the left. Save your file, share code with friends and open saved projects.

Want to change languages? Try the search bar up the top.

def find_substrings(input_string) -> | (char_1, index_1, length_1), (char_n, index_n, length_n)|

INPUT             OUTPUT
"aabcdeee"   ->    |("a", 0, 2), ("e", 5, 3)|
*/
import java.util.List;
import java.util.ArrayList;

/*
* Find string which are repeated characters along with character, index, count
* For e.g input string = "aabcdeee"
* return value as tuple of [('a',0,2), ('e',5,3)]
* */

public class FindSubStringAndReturnCharIndexCountAsTuple {

    public static class Tuple {
        private char character;
        private int index;
        private int count;


        public Tuple(char character, int index, int count) {
            this.character = character;
            this.index = index;
            this.count = count;
        }

        @Override
        public String toString() {
            return String.format("('%c', %d, %d)", character, index, count);
        }
    }
    public static List<Tuple> findSubStringCounter(String inputString) {
        if(inputString == null || inputString.isEmpty())
            throw new IllegalArgumentException("Pass Valid input");

        List<Tuple> result = new ArrayList<>();
        if(inputString.length() < 2)
            return result;

        int size = inputString.length();
        int start = 0;

        while(start < size) {
            char currentCharacter = inputString.charAt(start);
            int end = start;

            while(end < size && inputString.charAt(end) == currentCharacter) {
                end++;
            }

            int count = end - start;
            if(count > 1) {
                result.add(new Tuple(currentCharacter, start, count));
            }

            start = end;
        }
        return result;
    }

    public static void main(String args[]) {

        String inputString = "aabcdeee";

        List <Tuple> subsetString = findSubStringCounter(inputString);
        for(Tuple tuple : subsetString) {
            System.out.println(tuple);
        }
    }

}

